server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/MDNSendMethod.scala

```diff
@@ -0,0 +1,262 @@
+        .sendingMode(DispositionSendingMode.fromString(requestEntry.disposition.sendingMode).orElseThrow())
+        .build())
+      .finalRecipientField(requestEntry.finalRecipient.getOrElse(FinalRecipientField(sender)).value)
+
+    requestEntry.reportingUA.map(uaField => reportBuilder.reportingUserAgentField(ReportingUserAgent.builder().parse(uaField.value).build()))
+
+    requestEntry.extensionFields.map(extensions => extensions
+      .map(extension => reportBuilder.withExtensionField(ExtensionField.builder()
+        .fieldName(extension._1)
+        .rawValue(extension._2)
+        .build())))
+
+    val mdnBuilder = MDN.builder()
+      .report(reportBuilder.build())
+      .message(requestEntry.includeOriginalMessage
+        .filter(isInclude => isInclude.value)
+        .map(_ => getOriginalMessage(messageRelated)).toJava)
+
+    requestEntry.textBody.map(textBody => mdnBuilder.humanReadableText(textBody.value))
+
+    val newMessageId = messageIdFactory.generate().serialize()
+    val mdn = mdnBuilder.build()
+    val mimeMessage = mdn.asMimeMessage()
+
+    mimeMessage.setFrom(sender)
+    mimeMessage.setRecipients(javax.mail.Message.RecipientType.TO, getRecipientAddress(messageRelated, session))
+    mimeMessage.setHeader("Message-Id", newMessageId)
+    mimeMessage.setSubject(requestEntry.subject.getOrElse(SubjectField("subject todo")).value)
+
+    val mdnSendCreateResponse = buildMDNSendCreateResponse(requestEntry, mdn)
+    (MailImpl.fromMimeMessage(newMessageId, mimeMessage) -> mdnSendCreateResponse)
+  }
+
+  private def buildMDNSendCreateResponse(requestEntry: MDNSendCreateRequest, mdn: MDN) =
+    MDNSendCreateResponse(
+      subject = requestEntry.subject match {
+        case Some(_) => None
+        case None => Some(SubjectField(mdn.asMimeMessage().getSubject))
+      },
+      textBody = requestEntry.textBody match {
+        case Some(_) => None
+        case None => Some(TextBodyField(mdn.getHumanReadableText))
+      },
+      reportingUA = requestEntry.reportingUA match {
+        case Some(_) => None
+        case None => mdn.getReport.getReportingUserAgentField
+          .map(ua => ReportUAField(ua.fieldValue()))
+          .toScala
+      },
+      mdnGateway = mdn.getReport.getGatewayField
+        .map(gateway => MDNGatewayField(gateway.fieldValue()))
+        .toScala,
+      originalRecipient = mdn.getReport.getOriginalRecipientField
+        .map(originalRecipient => OriginalRecipientField(originalRecipient.fieldValue()))
+        .toScala,
+      includeOriginalMessage = requestEntry.includeOriginalMessage match {
+        case Some(_) => None
+        case None => Some(IncludeOriginalMessageField(mdn.getOriginalMessage.isPresent))
+      },
+      error = Option(mdn.getReport.getErrorFields.asScala
+        .map(error => ErrorField(error.getText.formatted()))
+        .toSeq)
+        .filter(error => error.nonEmpty),
+      extensionFields = requestEntry.extensionFields match {
```

Comment:
Why would we have extension field if the client did not specify it?

Reply:
I can't differentiate this field is `server-set`(server evaluate) or not.
So, This is my reserve.

Reply:
Here I think doing more Test Driven Development would be saving you quite a bunch of code...